pcdsutils/log.py

```diff
@@ -321,3 +338,256 @@ def centralized_logging_enabled() -> bool:
+    ----------
+    message : Warning
+        This is the Warning object created by a warnings.warn call.
+        When converted using str, this becomes the string message
+        that was passed into warnings.warn. This will be put into the
+        generated log message text and into the extra dict.
+    category : type[Warning]
+        The warning type, e.g. UserWarning, DeprecationWarning, etc.
+        this will be put into the generated log message text and into
+        the extra dict.
+    filename : str
+        The name of the source code file that generated the warning.
+        This will be put into the extra dict.
+    lineno : int
+        The line number in the file that generated the warning. This will
+        be put into the extra dict.
+    file : file-like, optional
+        A file-like object that is normally used in the warnings handler as
+        the destination for warnings, defaulting to sys.stderr. This will
+        be put into the extra dict.
+    line : str, optional
+        The string line in the file that generated the warning. I have never
+        seen this passed into the warning handler. This will be put into
+        the extra dict.
+    logger : Logger, optional
+        Use this argument to override the default logger for the warnings
+        handler, which is the warnings_logger defined in this module.
+        This is currently the pcdsutils.log.warnings logger.
+    """
+    logger.warning(
+        '%s: %s',
+        category.__name__,
+        message,
+        extra={
+            'warning_message': message,
+            'warning_category': category,
+            'warning_filename': filename,
+            'warning_lineno': lineno,
+            'warning_file': file,
+            'warning_line': line,
+        },
+    )
+
+
+def install_log_warning_handler(
+    logger: logging.Logger = warnings_logger,
+) -> None:
+    """
+    Replaces warnings.showwarning with the log_warning_handler above.
+
+    Parameters
+    ----------
+    logger : Logger, optional
+        Use this argument to override the default logger for the warnings
+        handler, which is the warnings_logger defined in this module.
+        This is currently the pcdsutils.log.warnings logger.
+    """
+    warnings.showwarning = functools.partial(
+        log_warning_handler,
+        logger=logger,
+    )
+
+
+def uninstall_log_warning_handler() -> None:
+    """
+    Restores the default behavior of the warnings module.
+
+    Intended to undo the effects of "install_log_warning_handler"
+    from this module.
+    """
+    warnings.showwarning = warnings._showwarning_orig
+
+
+@dataclasses.dataclass(eq=True, frozen=True)
+class WarningRecordInfo:
+    """
+    Hashable collection of the unique information from a warnings.warn call.
+    """
+    message: str
+    category: type[Warning]
+    filename: str
+    lineno: int
+
+    @staticmethod
+    def from_record(record: logging.LogRecord) -> "WarningRecordInfo":
```

Comment:
For future reference (I know we chatted about this briefly); the quotes can be circumvented with:
```python
from __future__ import annotations
```

Separately, it's not so important here/now, but a pattern I have tended to like recently is using a `classmethod` instead of `staticmethod` in scenarios like these. 

That way, if you have a custom `WarningRecordInfoSpecial` subclass, `WarningRecordInfoSpecial.from_record()` would generate the appropriate subclass instance instead of a `WarningRecordInfo`

Reply:
While I'm here, I'll `__future__` it up.

Also that's a good point, I didn't consider the consequences of using a static method for subclasses.